### DISCUSSION TITLE
`engine.begin()` tries to commit transaction on an invalidated connection

### DISCUSSION THREAD

**Discussion by agronholm** [sqlalchemy/sqlalchemy#12080] [Usage Questions] [answered]
When running a query in a task, and that task is cancelled, the underlying connection gets invalidated. That's all fine, but then the `engine.begin()` context manager tries to commit the transaction, which then raises a `PendingRollbackError`. Why is it not checking for an invalidated connection?

Here's a simple reproducer using `psycopg` against a local PostgreSQL database (`testdb`):
```py
import asyncio
from asyncio import CancelledError, create_task, sleep

from sqlalchemy import text
from sqlalchemy.ext.asyncio import create_async_engine


async def run_transaction():
    engine = create_async_engine("postgresql+psycopg://postgres:secret@localhost/testdb")
    async with engine.begin() as conn:
        task = create_task(conn.execute(text("SELECT PG_SLEEP(5)")))
        await sleep(0.5)
        task.cancel()
        try:
            await task
        except CancelledError:
            print("cancelled, in_transaction =", conn.in_transaction())
            print("cancelled, invalidated =", conn.invalidated)

asyncio.run(run_transaction())
```
I started investigating this because my ORM harness needs to check at exit if it should commit or not, and I have no clue how to do this with an `AsyncSession`:
```py
class MyORMHarness:
    def create_async_session(self) -> AsyncSession:
        async def teardown_session() -> None:
            try:
                if session.in_transaction():  # this returns True even if the connection has been invalidated!
                    if sys.exc_info()[1] is None:
                        await session.commit()
                    else:
                        await session.rollback()
            finally:
                await session.close()

        session: AsyncSession = self._async_sessionmaker()
        add_teardown_callback(teardown_session)
        return session
```
If I try to get the connection from the session, even the default bind, it raises an exception because it tries to revalidate the connection instead of just returning it. So there are two distinct issues here I think.

**Comment by CaselIT** [comment]
Hi,

For the first example I feel like the use case is invalid, since you are sharing the connection between two tasks. That's not supported.

The second use case looks a lot like you are also doing the same, but there aren't enough details in the example

**Reply by agronholm** [reply]
I'm not actually sharing a connection between multiple tasks – this was just something I did to show the effects of task cancellation on the connection and transaction. I'm not trying to run any queries in two tasks at the same time.

Is the connection somehow tied to a specific task, however (via contextvars or something)? That would indeed affect the results.

**Reply by zzzeek** [reply]
it's no different than if you just called conn.invalidate in a normal block

```py
import asyncio
from asyncio import CancelledError, create_task, sleep

from sqlalchemy import text
from sqlalchemy.ext.asyncio import create_async_engine


async def run_transaction():
    engine = create_async_engine("postgresql+psycopg://scott:tiger@localhost/test")
    async with engine.begin() as conn:
        await conn.invalidate()
        await conn.rollback()  # <--- you need to do this too if you invalidated

asyncio.run(run_transaction())
```

**Reply by CaselIT** [reply]
> I'm not actually sharing a connection between multiple tasks

well you are opening and closing it from one task and using it in another. that counts as sharing a connection between tasks.

I think that this case would not present itself if you opened, closed and used the collection all in a single task.

**Comment by zzzeek** [comment]
you can roll back the connection in the except block which ends the transaction.   "transaction" in SQLAlchemy is an abstraction that continues past "invalidated" since we want the try/except/finally blocks to be whole

```py
async def run_transaction():
    engine = create_async_engine("postgresql+psycopg://scott:tiger@localhost/test")
    async with engine.begin() as conn:
        task = create_task(conn.execute(text("SELECT PG_SLEEP(5)")))
        await sleep(0.5)
        task.cancel()
        try:
            await task
        except CancelledError:
            print("cancelled, in_transaction =", conn.in_transaction())
            print("cancelled, invalidated =", conn.invalidated)
            await conn.rollback()

```

**Reply by agronholm** [reply]
I think my issue is that somewhere down the line, an exception was raised and handled, without the transaction being explicitly rolled back. As this harness is a framework thing, should I just catch `PendingRollbackError` from `.commit()` and issue a rollback instead? The only other way is to ensure that every single exception bubbles to the harness, regardless of what third-party code sits in between. I'm not sure that's doable.

**Reply by zzzeek** [reply, marked as answer]
if you're calling .commit() explicitly, sure

**Reply by agronholm** [reply]
Alright, thanks, I think I understand this better now!
